In [3]:
from openai import OpenAI

/home/codespace/miniconda3/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


## Q1. Running Ollama with Docker

#### What's the version of ollama client?
- ollama version is 0.1.48

## Q2. Downloading an LLM

#### What's the content of the file related to gemma?
- {"schemaVersion":2,"mediaType":"application/vnd.docker.distribution.manifest.v2+json","config":{"mediaType":"application/vnd.docker.container.image.v1+json","digest":"sha256:887433b89a901c156f7e6944442f3c9e57f3c55d6ed52042cbb7303aea994290","size":483},"layers":[{"mediaType":"application/vnd.ollama.image.model","digest":"sha256:c1864a5eb19305c40519da12cc543519e48a0697ecd30e15d5ac228644957d12","size":1678447520},{"mediaType":"application/vnd.ollama.image.license","digest":"sha256:097a36493f718248845233af1d3fefe7a303f864fae13bc31a3a9704229378ca","size":8433},{"mediaType":"application/vnd.ollama.image.template","digest":"sha256:109037bec39c0becc8221222ae23557559bc594290945a2c4221ab4f303b8871","size":136},{"mediaType":"application/vnd.ollama.image.params","digest":"sha256:22a838ceb7fb22755a3b0ae9b4eadde629d19be1f651f73efb8c6b4e2cd0eea0","size":84}]}

## Q3. Running the LLM

#### Test the following prompt: "10 * 10". What's the answer?
- "10 * 10 = 100"

## Q4. Downloading the weights

#### What's the size of the ollama_files/models folder?
- 1.6G 

## Q5. Adding the weights

#### What do you put after COPY?
- ./ollama_files

## Q6. Serving it

In [19]:
client = OpenAI(
    base_url='http://localhost:11434/v1/',
    api_key='ollama',
)

In [33]:
def llm(prompt):
    response = client.chat.completions.create(
        temperature=0.0,
        model='gemma:2b',
        messages=[{"role": "user", "content": prompt}]
    )
    
    return response

In [ ]:
prompt = "What's the formula for energy?"
response = llm(prompt)

In [44]:
response.to_dict()

{'id': 'chatcmpl-630',
 'choices': [{'finish_reason': 'stop',
   'index': 0,
   'message': {'content': "Sure, here's the formula for energy:\n\n**E = K + U**\n\nWhere:\n\n* **E** is the energy in joules (J)\n* **K** is the kinetic energy in joules (J)\n* **U** is the potential energy in joules (J)\n\n**Kinetic energy (K)** is the energy an object possesses when it moves or is in motion. It is calculated as half the product of an object's mass (m) and its velocity (v) squared:\n\n**K = 1/2 * m * v^2**\n\n**Potential energy (U)** is the energy an object possesses when it is in a position or has a specific configuration. It is calculated as the product of an object's mass and the gravitational constant (g) multiplied by the height or distance of the object from a reference point.\n\n**Gravitational potential energy (U)** is given by the formula:\n\n**U = mgh**\n\nWhere:\n\n* **m** is the mass of the object in kilograms (kg)\n* **g** is the acceleration due to gravity in meters per second 

In [52]:
response.usage.completion_tokens

304

#### How many completion tokens did you get in response?
- 304